
# This notebook has codes for random forest from scikit learn<br> and applying the bagging method with voting using scikit learn decision tree for a data science class assignment.



# Part A


In [114]:

import numpy as np

import pandas as pd


cancer_df  = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Cancer.csv")

cancer_df.head()



Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
0                5                        1                         1   
1                5                        4                         4   
2                3                        1                         1   
3                6                        8                         8   
4                4                        1                         1   

   Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
0                  1                            2            1   
1                  5                            7           10   
2                  1                            2            2   
3                  1                            3            4   
4                  3                            2            1   

   Bland_Chromatin  Normal_Nucleoli  Mitoses  Malignant_Cancer  
0                3                1        1                 0  
1                3                2        1                 0  
2                3                1        1                 0  
3                3                7        1                 0  
4                3                1        1                 0


# Part B


In [115]:



from sklearn.model_selection import train_test_split

# get all the columns

cancer_feature_columns = list( cancer_df.columns )

# remove the column that is the label

cancer_feature_columns.remove ('Malignant_Cancer')

# make the feature matrix

Cancer_Feature_Matrix = cancer_df[cancer_feature_columns]


Cancer_Label_Vector = cancer_df['Malignant_Cancer']

# creating the dataset splits
CFM_Training, CFM_Testing, CLV_Training, CLV_Testing = train_test_split(Cancer_Feature_Matrix,Cancer_Label_Vector, test_size = 0.35, random_state = 3)



# Part C


In [116]:

from sklearn.tree import DecisionTreeClassifier

from sklearn.metrics import accuracy_score

d_tree_state_value = 3

dec_tree_classifier = DecisionTreeClassifier(random_state = d_tree_state_value )

dec_tree_classifier.fit(CFM_Training,CLV_Training)

dec_tree_pred = dec_tree_classifier.predict(CFM_Testing)

dec_tree_acc = accuracy_score(CLV_Testing,dec_tree_pred)

print (dec_tree_acc)

0.8301886792452831



# Part C Results
At this time, the accuracy score is calculated to be: <br>
0.8301886792452831



# Part D


In [117]:

# needed this
from sklearn.utils import resample

# needed this
from collections import Counter



size_of_original_dataset = len(cancer_df.index)

bootstrap_size = int(0.8 * size_of_original_dataset)

dec_tree_predict_list = []

for i in range (0, 19, 1):
    
    # resampling the features training set
    CFM_BS_DS_Training = resample(CFM_Training, n_samples = bootstrap_size, random_state = i, replace = True)
    
    # resampling the labels training set
    CLV_BS_DS_Training = resample(CLV_Training, n_samples = bootstrap_size, random_state = i, replace = True)
    
    base_decision_tree = DecisionTreeClassifier(random_state = 3)
    
    base_decision_tree.fit(CFM_BS_DS_Training, CLV_BS_DS_Training)
    
    current_dec_tree_predict = base_decision_tree.predict(CFM_Testing)
    
    dec_tree_predict_list.append(current_dec_tree_predict)


# I made a dataframe to check the data
# I used the features inside pandas dataframe to
# create rows from the columns of the dataframe
result_df = pd.DataFrame(data = dec_tree_predict_list)    

dec_trees_predictions_list = []

dec_trees_final_prediction_list = []

# this loop makes a list with it's rows as the columns of the
# result dataframe
for current_column in result_df.columns:
    
    current_column_list = result_df[current_column].tolist()
    dec_trees_predictions_list.append(current_column_list)
    
# I use Counter from Collections to  do voting
# it's recommended for more than 1 object counting
for current_row in dec_trees_predictions_list:
    
    current_row_counter = Counter (current_row)
    
    if current_row_counter[1] > current_row_counter[0]:
        dec_trees_final_prediction_list.append(1)
        
    elif current_row_counter[0] > current_row_counter[1]:
        dec_trees_final_prediction_list.append(0)
        
    else:
        dec_trees_final_prediction_list.append(0)
        
accuracy_rate = accuracy_score (CLV_Testing, dec_trees_final_prediction_list)
    
print(accuracy_rate)

0.8867924528301887



# Part D Results
Bases on the accuracy soore of the ensemble. <br>
At this time the accuracy score is calculated to be: 0.8867924528301887. <br>
This seems to be an improvement compared to the single decision tree's <br>
accuracy results from above.



# Part E


In [118]:

import numpy as np

import pandas as pd

from sklearn.model_selection import train_test_split

from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import accuracy_score


cancer_df  = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Cancer.csv")

cancer_feature_columns = list( cancer_df.columns )

cancer_feature_columns.remove ('Malignant_Cancer')

Cancer_Feature_Matrix = cancer_df[cancer_feature_columns]

Cancer_Label_Vector = cancer_df['Malignant_Cancer']

CFM_Training, CFM_Testing, CLV_Training, CLV_Testing = train_test_split(Cancer_Feature_Matrix,Cancer_Label_Vector, test_size = 0.35, random_state = 3)


random_forest_classifier = RandomForestClassifier(n_estimators = 19, bootstrap = True, random_state=3)

random_forest_classifier.fit(CFM_Training,CLV_Training)

rand_forest_prediction = random_forest_classifier.predict(CFM_Testing)

rand_forest_acc_score = accuracy_score(CLV_Testing,rand_forest_prediction)

print (rand_forest_acc_score)



0.9245283018867925



# Part E Result
At the time of calculation, the accuracy of the <br>
random forest classifier with this data is: 0.9245283018867925.

